server/ft.go

```diff
@@ -0,0 +1,187 @@
+		// If there is an error, we have to stop now.
+		if err != nil {
+			node.Close()
+			return fmt.Errorf("ft: error getting the store lock: %v", err)
+		}
+		// If ok is false, it means that we did not get the lock
+		// so we should be going back to standby.
+		Noticef("STAN: ft: unable to get store lock at this time, going back to standby")
+		node.Close()
+		goto graft_create_node
+	}
+	// Server could have been shutdown
+	s.mu.Lock()
+	if s.shutdown {
+		s.mu.Unlock()
+		return nil
+	}
+	s.wg.Add(1)
+	s.mu.Unlock()
+	Noticef("STAN: Server confirmed leader")
+	go func() {
+		// On shutdown, server is waiting on s.wg, so release when
+		// we exit this function.
+		defer s.wg.Done()
+		// We are going to care only about state changes here, and not the
+		// error channel. The reason is that for a leader, the only case
+		// where we would get an error is writting the state, which would
+		// happen if the leader gets demoted, which we handle in the state
+		// change case. So we will simply dequeue the error channel and
+		// print, but nothing else.
+		// Note that when we call "Fatalf", it is possible that the server
+		// was started programmatically, and that no logger is set, in which
+		// case "Fatalf" does nothing, so each "Fatalf" call is followed by
+		// a panic.
+		for {
+			select {
+			case sc := <-stateChangeChan:
+				if sc.To != graft.LEADER {
+					err = fmt.Errorf("ft: server demoted to %v, aborting", sc.To.String())
+					Fatalf("STAN: %v", err)
+					if noPanic {
+						node.Close()
+						s.setFTError(err)
+					} else {
+						panic(err)
+					}
+				}
+				Noticef("STAN: ft: state change, from %v to %v", sc.From.String(), sc.To.String())
+			case err := <-errChan:
+				Errorf("STAN: ft: error: %v", err)
+			case <-s.ftQuit:
+				node.Close()
+				return
+			}
+		}
+	}()
+	// Start the recovery process, etc..
+	return s.start(FTActive)
+}
+
+// ftCreateRAFTNode creates a RAFT node and returns required channels.
+// Since there is no step down feature, if a node is elected leader but
+// the lock file checks shows that there is another leader currently running,
+// the server will close and create a new node. State and error channels
+// (handler) could be reused, but creating new ones to avoid possible updates
+// from previous instance to interfere with new one.
+func (s *StanServer) ftCreateRAFTNode(opts *nats.Options) (*graft.Node, chan graft.StateChange, chan error, error) {
+	// Loop until we can get a valid connection...
+	var (
+		err error
+		rpc *graft.NatsRpcDriver
+	)
+	// We need to have a NATS RPC in order for RAFT to work,
+	// so try until we succeed.
+	printCount := 0
+	for {
+		rpc, err = graft.NewNatsRpc(opts)
+		if err != nil {
+			if printCount%10 == 0 {
+				Errorf("STAN: ft: creating RPC failed (%v), trying again...", err)
+			}
+			printCount++
+			select {
+			case <-s.ftQuit:
+				return nil, nil, nil, nil
+			case <-time.After(500 * time.Millisecond):
```

Comment:
RAFT will not work without the RPC in place, so I want to try as often as possible - not too crazy - but not delay more than needed.